beacon-chain/core/state/transition.go

```diff
@@ -404,41 +420,46 @@ func ProcessBlockNoVerifyAnySig(
 	ctx, span := trace.StartSpan(ctx, "beacon-chain.ChainService.state.ProcessBlockNoVerifyAnySig")
 	defer span.End()
 
-	state, err := b.ProcessBlockHeaderNoVerify(state, signed.Block)
-	if err != nil {
-		traceutil.AnnotateError(span, err)
-		return nil, nil, errors.Wrap(err, "could not process block header")
-	}
-	bSet, err := b.BlockSignatureSet(state, signed)
-	if err != nil {
-		traceutil.AnnotateError(span, err)
-		return nil, nil, errors.Wrap(err, "could not retrieve block signature set")
-	}
-	rSet, err := b.RandaoSignatureSet(state, signed.Block.Body)
-	if err != nil {
-		traceutil.AnnotateError(span, err)
-		return nil, nil, errors.Wrap(err, "could not retrieve randao signature set")
-	}
-	state, err = b.ProcessRandaoNoVerify(state, signed.Block.Body)
-	if err != nil {
-		traceutil.AnnotateError(span, err)
-		return nil, nil, errors.Wrap(err, "could not verify and process randao")
-	}
-
-	state, err = b.ProcessEth1DataInBlock(ctx, state, signed)
-	if err != nil {
-		traceutil.AnnotateError(span, err)
-		return nil, nil, errors.Wrap(err, "could not process eth1 data")
+	var bSet *bls.SignatureSet
+	var rSet *bls.SignatureSet
+	var aSet *bls.SignatureSet
+	var pipeline = []processFunc{
+		b.ProcessBlockHeaderNoVerify,
+		func(ctx context.Context,
+			state *stateTrie.BeaconState,
+			signed *ethpb.SignedBeaconBlock,
+		) (*stateTrie.BeaconState, error) {
+			var err error
+			bSet, err = b.BlockSignatureSet(state, signed)
+			return state, err
+		},
+		func(ctx context.Context,
+			state *stateTrie.BeaconState,
+			signed *ethpb.SignedBeaconBlock,
+		) (*stateTrie.BeaconState, error) {
+			var err error
+			rSet, err = b.RandaoSignatureSet(state, signed.Block.Body)
+			return state, err
+		},
+		b.ProcessRandaoNoVerify,
+		b.ProcessEth1DataInBlock,
+		ProcessOperationsNoVerifyAttsSigs,
+		func(ctx context.Context,
+			state *stateTrie.BeaconState,
+			signed *ethpb.SignedBeaconBlock,
+		) (*stateTrie.BeaconState, error) {
+			var err error
+			aSet, err = b.AttestationSignatureSet(ctx, state, signed.Block.Body.Attestations)
+			return state, err
+		},
```

Comment:
Not the most elegant code, but I couldn't think of a better option to incorporate signature functions into the pipeline. To make the pipeline a package-level variable, it is necessary to move signature variables to package-level as well. I find this scoping clearer, but I don't have a strong opinion.

Reply:
I think this is worse than before. For that particular reason, I'm more inclined to leave `ProcessBlockNoVerifyAnySig` as it is. Having good readability out-weights the original intentions in my opinion...